### PR TITLE
Fix: Extra information about partial updates in responses #3499

### PR DESCRIPTION
**Pull request by Anjali-NEC**
Fix issue #3499

### CODE REVIEW

File: src/lib/apiTypesV2/Attribute.cpp [hunk 1]
```diff
@@ -125,6 +125,7 @@ void Attribute::fill(const QueryContextResponse& qcrs, const std::string& attrNa
     contextAttributeP = NULL;
     // Look for the attribute by name
 
+    std::string NotExistingAtrr;
```
Comment (fgalan) [inline src/lib/apiTypesV2/Attribute.cpp:128]:
```suggestion
    std::string notExistingAtrr;
```

as variables use camelCase

Comment (Anjali-NEC) [reply src/lib/apiTypesV2/Attribute.cpp]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4097/commits/2dbb2ed15aee6867d16807ce665a18c7bf282c50

**Comment by fgalan** [comment]
Looking to the .test that has been modified, it seems you have added the extra information to response to `GET /v2/entities/{id}/attrs/{name}` and `GET /v2/entities/{id}/attrs/{name}/value`. However, note that is not the purpose of the issue. The issue refers to *partial updates* not to queries.

I'd suggest to model the case in a .test, roughly this way:

1. Create entity E with attribute A=1
2. Do a strict update E A=10 B=20 C=30, get error about entity does not have attribute B and C
3. Get E, check A=10
4. Do a strict update E A=100 B=200, get error about entity does not have attribute B
5. Get E, check A=100

and use that .test to guide your implementation

**Comment by fgalan** [comment]
> Looking to the .test that has been modified, it seems you have added the extra information to response to `GET /v2/entities/{id}/attrs/{name}` and `GET /v2/entities/{id}/attrs/{name}/value`

As an additional side note, note that telling the user `"The entity does not have such an attribute: { E1-T1 : [NONE] }"` is not very useful for the user in the case of `GET /v2/entities/E1/attrs/NONE` as the user already knows she has asked for `NONE` attribute ;)

**Comment by Anjali-NEC** [comment]
> I'd suggest to model the case in a .test, roughly this way:
> 
> 1. Create entity E with attribute A=1
> 2. Do a strict update E A=10 B=20 C=30, get error about entity does not have attribute B and C
> 3. Get E, check A=10
> 4. Do a strict update E A=100 B=200, get error about entity does not have attribute B
> 5. Get E, check A=100
> 
> and use that .test to guide your implementation

As per your suggestion I tried to run the above test and in second and forth step I got response "one or more of the attributes in the request already exist: [ A ]" instead of "entity does not have attribute B and C".

Below is the test case I am using:
```
echo "01. Create entity E with attribute A=1"
echo "======================================"
payload='{
  "id": "E",
  "type": "T",
  "A": {
    "value": 1
  }
}'
orionCurl --url /v2/entities --payload "$payload"
echo
echo


echo "02. Strict update E A=10 B=20 C=30, get error about entity does not have attribute B and C"
echo "=========================================================================================="
payload='{
  "actionType": "appendStrict",
  "entities": [
    {
      "id": "E",
      "type": "T",
      "A": {
        "value": 10
      },
      "B": {
        "value": 20
      },
      "C": {
        "value": 30
      }
    }
  ]
}'
orionCurl --url /v2/op/update --payload "$payload"
echo
echo

echo "03. Get E, check A=10"
echo "====================="
orionCurl --url /v2/entities/E
echo
echo

echo "04. Strict update E A=100 B=200 C=30, get error about entity does not have attribute B"
echo "======================================================================================"
payload='{
  "actionType": "appendStrict",
  "entities": [
    {
      "id": "E",
      "type": "T",
      "A": {
        "value": 100
      },
      "B": {
        "value": 200
      }
    }
  ]
}'
orionCurl --url /v2/op/update --payload "$payload"
echo
echo


echo "05. GET E, check A=100"
echo "======================"
orionCurl --url /v2/entities/E
echo
echo

OUTPUT
01. Create entity E with attribute A=1
======================================
HTTP/1.1 201 Created
Content-Length: 0
Location: /v2/entities/E?type=T
Fiware-Correlator: 16bfee22-cfb0-11ec-b47d-f9ebd88f9903
Date: Mon, 09 May 2022 15:52:54 GMT



02. Strict update E A=10 B=20 C=30, get error about entity does not have attribute B and C
==========================================================================================
HTTP/1.1 422 Unprocessable Entity
Content-Length: 107
Content-Type: application/json
Fiware-Correlator: 16c39996-cfb0-11ec-b47d-f9ebd88f9903
Date: Mon, 09 May 2022 15:52:54 GMT

{
    "description": "one or more of the attributes in the request already exist: [ A ]",
    "error": "Unprocessable"
}


03. Get E, check A=10
=====================
HTTP/1.1 200 OK
Content-Length: 161
Content-Type: application/json
Fiware-Correlator: 16c912ae-cfb0-11ec-b47d-f9ebd88f9903
Date: Mon, 09 May 2022 15:52:54 GMT

{
    "A": {
        "metadata": {},
        "type": "Number",
        "value": 1
    },
    "B": {
        "metadata": {},
        "type": "Number",
        "value": 20
    },
    "C": {
        "metadata": {},
        "type": "Number",
        "value": 30
    },
    "id": "E",
    "type": "T"
}


04. Strict update E A=100 B=200 C=30, get error about entity does not have attribute B
======================================================================================
HTTP/1.1 422 Unprocessable Entity
Content-Length: 110
Content-Type: application/json
Fiware-Correlator: 16cd8546-cfb0-11ec-b47d-f9ebd88f9903
Date: Mon, 09 May 2022 15:52:54 GMT

{
    "description": "one or more of the attributes in the request already exist: [ A, B ]",
    "error": "Unprocessable"
}


05. GET E, check A=100
======================
HTTP/1.1 200 OK
Content-Length: 161
Content-Type: application/json
Fiware-Correlator: 16d1f2b6-cfb0-11ec-b47d-f9ebd88f9903
Date: Mon, 09 May 2022 15:52:54 GMT
{
    "A": {
        "metadata": {},
        "type": "Number",
        "value": 1
    },
    "B": {
        "metadata": {},
        "type": "Number",
        "value": 20
    },
    "C": {
        "metadata": {},
        "type": "Number",
        "value": 30
    },
    "id": "E",
    "type": "T"
}

```

Please let me know what changes I need to be done in the test case for expected response.

**Comment by fgalan** [comment]
Sorry for the delay answering you...

By "strict update" y mean

```
"actionType": "update",

(Note that "appendStrict" refers to "strict append", which is not the same thant "strict update")

Change that in your .test and let's wee how it goes...

**Comment by Anjali-NEC** [comment]
> Sorry for the delay answering you...
> 
> By "strict update" y mean
> 
> ```
> "actionType": "update",
> ```
> 
> (Note that "appendStrict" refers to "strict append", which is not the same thant "strict update")
> 
> Change that in your .test and let's wee how it goes...

After so long waiting for your reply I have started working on another issue. Currently I am little busy on another activity will work on that after in future.

**Comment by Chandradeep-NEC** [comment]
@fgalan 
We apology for putting the issue on hold for long period. Anjali-NEC was working on other high priority issue (i.e. #4149). Currently she is on long holiday due to her personal reasons.
Now we have planned to resume working on this ticket in mid of February.

**Comment by fgalan** [comment]
Overpassed by PR #4447 

@Anjali-NEC @Chandradeep-NEC I'd recommend you to have a look to that PR for learning purposes, to check how this issue (#3499) can be solved.

**Pull request closed without merging**